### PR TITLE
Allow input simulation service to work outside of editor

### PR DESCRIPTION
**Pull request by Troy-Ferrell**
## Overview
This change allow for the input simulation service to be used outside of editor play mode such as in standalone builds. 

To enable this, the following was done
- Moved JSON articulated hand pose data into SImulatedArticulatedHandPoses.cs file as const string to avoid MRTKFiles loading and resouces loading dependencies
- Moved InputSimulationService and InputPlaybackService from editor asdmef to InputSimulation asmdef
- Created migration code path in SystemType.cs to handle InputSimulationService reference values stored in MRTK profiles (such as MRTKInputSystemProfile) to new destination (i.e non-editor asmdef)
- Reserialized all profile assets in MRTK repo
- "prettied" JSON files with articulated hand poses
- Added menu item for Reserialize Selection with other similar entries
- Made InputSimulation.asmdef target all platforms

**NOTE**
This new feature capability is not enabled by default. The default profiles in MRTK still have the runtime platform setting for input simulation service to not include windows standalone. This is to ensure people aren't broken for those that don't want this as keys and mouse input will route to the simulation service.

image

## Changes
- Fixes: #3692 #6966

### CODE REVIEW

**Comment by Troy-Ferrell** [comment]
@thalbern re-based and updated documentation

**Comment by julenka** [comment]
Out of curiosity, is there a particular example use case you are enabling here, like somebody wants to actually use hand simulation in a standalone build?

**Comment by Troy-Ferrell** [comment]
@julenka yes, see details in related issues: Fixes: #3692 #6966

Particularly the latter issue has been up-voted by multiple parties 5 times

File: Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs [hunk 1]
```diff
@@ -30,6 +30,7 @@ private static void ReserializeMaterials()
             AssetDatabase.ForceReserializeAssets(array);
         }
 
+        [MenuItem("Mixed Reality Toolkit/Utilities/Reserialize/Reserialize Selection")]
         [MenuItem("Assets/Mixed Reality Toolkit/Reserialize Selection")]
```
[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs]: No, one is a right click of the mouse and the other is from the top tab bar

File: Assets/MixedRealityToolkit/Definitions/Utilities/SystemType.cs [hunk 2]
```diff
@@ -73,6 +74,10 @@ public SystemType(Type type)
 
         void ISerializationCallbackReceiver.OnAfterDeserialize()
         {
+            // Class references may move between asmdef or be renamed throughout MRTK development
+            // Check to see if we need to update our reference value
+            reference = TryMigrateReference(reference);
```
[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Definitions/Utilities/SystemType.cs]: hmmm good call, I'll look into it

**Comment by julenka** [comment]
>Out of curiosity, is there a particular example use case you are enabling here, like somebody wants to actually use hand simulation in a standalone build?

Answering my own question: I actually used the code in this PR for a personal project this weekend :-)

File: Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs [hunk 3]
```diff
@@ -211,17 +211,13 @@ public override void Initialize()
         {
             base.Initialize();
 
-            ArticulatedHandPose.LoadGesturePoses();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs]: Since the JSON files are moving to classes, how will this impact ability to have custom input simulation gestures?

File: Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs [hunk 4]
```diff
@@ -30,6 +30,7 @@ private static void ReserializeMaterials()
             AssetDatabase.ForceReserializeAssets(array);
         }
 
+        [MenuItem("Mixed Reality Toolkit/Utilities/Reserialize/Reserialize Selection")]
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs]: Maybe put this in separate PR? Seems unrelated to this change

File: Documentation/InputSimulation/InputSimulationService.md [hunk 5]
```diff
@@ -69,58 +76,58 @@ The input simulation supports emulated hand devices. These virtual hands can int
...
+Press and hold the **Left/Right Hand Control Key** (default: *Left Shift* for left hand and *Space* for right hand) to gain control of either hand. While the manipulation key is pressed, the hand will appear in the viewport. Once the manipulation key is released, the hands will disappear after a short **Hand Hide Timeout**.
 
-Hands can be toggled on permanently in the [input simulation tools window](#input-simulation-tools-window) or by pressing the __Toggle Left/Right Hand Key__ (default: *T* for left and *Y* for right). Press the toggle key again to hide the hands again.
+Hands can be toggled on permanently in the [input simulation tools window](#input-simulation-tools-window) or by pressing the **Toggle Left/Right Hand Key** (default: *T* for left and *Y* for right). Press the toggle key again to hide the hands again.
 
-Mouse movement will move the hand in the view plane. Hands can be moved further or closer to the camera using the __mouse wheel__.
+Mouse movement will move the hand in the view plane. Hands can be moved further or closer to the camera using the **mouse wheel**.
```
Comment (julenka) [inline Documentation/InputSimulation/InputSimulationService.md]: Aside: I wish this was customizable, like I could press ctrl + up to move forward, or 'i' to move forward the hand. It's hard to scroll on my mac :-)

File: Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs (hunk 4 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tools/ReserializeAssetsUtility/ReserializeAssetsUtility.cs]: I guess it's not necessary but seems severe to remove from this PR and create a whole separate one for it and issue item

File: Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs (hunk 3 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs]: hmmm, I mean the original implementation was hardcoded as well. 

But it is a good point to consider the flexibility of this new design. One could argue that the SimulatedArticulatedHandposes are a custom cache implementation and thus someone could create their own just like it...they would just have to update the other service implementations as well.

I could look at adding accessors to SimulatedArticulatedHandposes

File: Documentation/InputSimulation/InputSimulationService.md (hunk 5 above)
Comment (Troy-Ferrell) [reply Documentation/InputSimulation/InputSimulationService.md]: 100. Especially it was annoying for keyboards that aren't US based. You can re-route input keys in Unity I believe and the input simulation profile but I can't recall how extensive it is

**Comment by Troy-Ferrell** [comment]
> find

I don't believe so but I'll look into that. I believe you are talking about the InputPlaybackService? Not sure if we could record custom gestures and save them to JSON but let's see

File: Assets/MixedRealityToolkit/Definitions/Utilities/SystemType.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Definitions/Utilities/SystemType.cs]: Yea I am going to keep these separate for now. The Migration is "we know this has changed so let us fix it". The TryRepair is "hey we can attempt to fix this, but we aren't sure which one is correct". 

I also don't want the type filtering/name checking code to execute for every system type deserialization. We would need to create a type cache so we don't have terrible performance impacts every time we load the project....all when we know the vast majority of type names don't need to be checked.

File: Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs (hunk 3 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Services/InputSimulation/InputSimulationService.cs]: Added setter to override
